python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+
+
+class KuberayProvider(ICloudInstanceProvider):
+    """
+    This class is a thin wrapper around the Kuberay API client. It modifies
+    the RayCluster resource spec on the Kuberay API server to scale the cluster:
+
+    It launches new instances/nodes by submitting a scale request to the Kuberay API
+    to overwrite "replicas" fields in the RayCluster resource on the Kubernetes API
+    server.
+
+    It terminates instances/nodes by submitting a scale request to the Kuberay API to
+    overwrite "workersToDelete" fields in the RayCluster resource on the Kubernetes
+    API server, as well as adjusting the "replicas" fields in the RayCluster resource
+    spec.
+
+    """
+
+    def __init__(
+        self,
+        cluster_name: str,
+        namespace: str,
+        k8s_api_client: IKubernetesHttpApiClient,
+    ):
+        """
+        Args:
+            cluster_name: The name of the RayCluster resource.
+            namespace: The namespace of the RayCluster resource.
+            k8s_api_client: The client to the Kuberay API server.
+        """
+        self._cluster_name = cluster_name
+        self._namespace = namespace
+
+        self._k8s_api_client = k8s_api_client
+
+        # Below are states that are cached locally.
+        self._requests = set()
+        self._launch_errors_queue = []
+        self._terminate_errors_queue = []
+
+        # Below are states that are fetched from the Kuberay API server.
+        self._ray_cluster = None
+        self._cached_instances: Dict[CloudInstanceId, CloudInstance]
+
+    @dataclass
+    class ScaleRequest:
+        """Represents a scale request that contains the current states and go-to states
+        for the ray cluster.
+
+        This class will be converted to patches to be submitted to the Kuberay API
+        server.
+        """
+
+        # The desired number of workers for each node type.
+        desired_num_workers: Dict[NodeType, int] = field(default_factory=dict)
+        # The workers to delete for each node type.
+        workers_to_delete: Dict[NodeType, List[CloudInstanceId]] = field(
+            default_factory=dict
+        )
+        # The worker groups to clear the workersToDelete field.
+        worker_groups_to_clear_delete: Set[NodeType] = field(default_factory=set)
+        # The worker groups that still have workers to be deleted.
+        worker_groups_with_pending_deletes: Set[NodeType] = field(default_factory=set)
+
+    ################################
+    # Interface for ICloudInstanceProvider
+    ################################
+
+    def get_non_terminated(self) -> Dict[CloudInstanceId, CloudInstance]:
+        self._sync_with_api_server()
+        return copy.deepcopy(
+            {
+                id: instance
+                for id, instance in self._cached_instances.items()
+                if instance.status != NodeStatus.TERMINATED
+            }
+        )
+
+    def terminate(self, ids: List[CloudInstanceId], request_id: str) -> None:
```

Comment:
Just for my education, who calls the `terminate` and `launch` functions?

Reply:
instance manager (and its subscribers)